# Pandas 과제 · 다음 축제에는 어떤 프로그램을 늘릴까?

학번: 2021041005　이름: 김태영

축제 신청 자료를 정리하고, 프로그램별로 몇 명이 신청했으며 실제로 몇 명이 참여했는지 알아보는 과제입니다. 마지막에는 계산 결과를 근거로 다음 축제에서 늘리고 싶은 프로그램을 하나 제안합니다.

## 과제 진행 방법

- 총 5문제입니다. 수업에서 배운 기능만 사용합니다. 그래프나 머신러닝은 필요하지 않습니다.
- 각 코드 지시문 **바로 아래의 코드셀**에 풀이를 작성하고 실행합니다. 결과는 `print()`로 출력하세요.
- 각 문제의 코드 풀이가 끝나면 결과를 해석하는 질문이 나옵니다. **질문 바로 아래의 답변 셀**을 더블클릭해 글을 적고, Shift+Enter로 마무리하세요.
- Q01은 원자료 `raw`, Q02~Q04는 작업용 표 `clean`, Q05는 Q04에서 만든 요약표 `summary`를 사용합니다. 위에서부터 순서대로 풀어 주세요.
- 제출 파일: 코드·출력 결과·답변을 담은 `Pandas_과제_학번_이름.ipynb` 1개.

## 데이터는 어떻게 구성되어 있나요?

`data/campus_festival.csv` 파일을 사용하여 문제를 풀어주세요. 조건은 다음과 같습니다.

- 학생들은 축제에 참여하기 위해 프로그램을 신청하였고, 실제 참여 여부 및 그 외 정보를 수집한 자료입니다.
- 학생 한 명은 한 프로그램에만 신청했습니다.
- 원래는 학생 한 명당 신청 기록이 한 행씩 있어야 합니다. 그러나 저장 과정에서 일부 행이 똑같이 여러 번 들어갔습니다.
- 같은 학생의 반복 기록은 신청번호를 포함한 모든 열의 내용이 같습니다.
- 일부 학생은 이동 시간이나 프로그램을 알게 된 경로를 응답하지 않았습니다. 이 빈칸이 **결측값**입니다.

| 열 이름 | 뜻 | 값의 예와 설명 |
|---|---|---|
| RegistrationId | 신청번호 | 서로 다른 학생은 다른 번호를 가집니다. |
| Program | 신청한 프로그램 | Concert: 공연 / EscapeRoom: 방탈출 / Maker: 만들기 체험 / Game: 게임 대회 |
| Year | 학년 | 1, 2, 3, 4 |
| TravelMinutes | 예상 이동 시간 | 집에서 축제 장소까지 걸리는 편도 시간(분). 신청할 때 응답한 값이며 일부 빈칸이 있습니다. |
| Channel | 프로그램을 알게 된 경로 | SNS: 소셜미디어 / Friend: 친구 / Poster: 포스터. 일부 빈칸이 있습니다. |
| Attended | 실제 참여 여부 | 소문자 o: 참여 / x: 미참여. 문자 o는 숫자 0과 다릅니다. 이 열에는 빈칸이 없습니다. |

## 시작 코드

압축파일을 풀고 이 노트북과 `data` 폴더를 같은 폴더에 둡니다. 아래 셀로 원자료를 읽은 후 Q01부터 시작하세요.
파일을 찾지 못하면 `data_path`를 실제 CSV 파일 경로로 바꾸세요.

In [7]:
import pandas as pd

data_path = "./campus_festival.csv"
raw = pd.read_csv(data_path)
print(raw.head())

   RegistrationId     Program  Year  TravelMinutes Channel Attended
0            1001  EscapeRoom     2           10.0     SNS        o
1            1002       Maker     3           15.0  Poster        o
2            1003        Game     1           40.0  Poster        x
3            1004     Concert     3            NaN  Friend        o
4            1005  EscapeRoom     4           15.0     SNS        o


### Q01-1 · 표의 크기 확인

`raw`의 행 수와 열 수를 출력하세요.

사용할 수 있는 기능: `shape`

In [2]:
# 여기에 코드를 작성하세요.
num_row = raw.shape[0]
num_col = raw.shape[1]

print("num_row : ",num_row,"\nnum_col : ",num_col)

num_row :  205 
num_col :  6


### Q01-2 · 빈칸 개수 확인

`raw`의 각 열에 빈칸이 몇 개 있는지 출력하세요. 빈칸이 없는 열은 0개로 표시해도 됩니다.

사용할 수 있는 기능: `isna()`, `sum()`

In [5]:
# 여기에 코드를 작성하세요.
for attribute in raw.columns:
    print(attribute," : ",raw[attribute].isna().sum())

RegistrationId  :  0
Program  :  0
Year  :  0
TravelMinutes  :  22
Channel  :  14
Attended  :  0


### Q01-3 · 추가로 반복된 행 수 확인

완전히 같은 행이 여러 번 있으면, 첫 번째 행을 제외하고 **추가로 반복된 행이 총 몇 개인지** 출력하세요.

예를 들어 한 기록이 3번 저장되어 있으면 추가 반복은 2행입니다. 다른 기록도 반복되었다면 각각의 반복 행 수를 합합니다.

‘같은 행’은 **두 행의 각 열을 비교했을 때 내용이 모두 일치한다**는 뜻입니다. 한 행 안의 값들이 서로 같다는 뜻은 아닙니다.

사용할 수 있는 기능: `duplicated()`, `sum()`

In [6]:
# 여기에 코드를 작성하세요.
print(raw.duplicated().sum())

5


**Q01 · 결과 질문 1**

Q01-2의 출력에서 빈칸이 있는 열은 무엇인가요? 열 이름과 각각의 빈칸 개수를 적으세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

TravelMinutes  :  22

Channel  :  14
<br><br>

**Q01 · 결과 질문 2**

Q01-3에서 반복된 행이 발견되었나요? 그 결과를 근거로, **현재 raw의 행 수를 그대로 학생 수라고 말해도 되는지** 한 문장으로 설명하세요. 학생 한 명당 원래 신청 기록은 한 행이라는 조건을 이용하세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

아니오.

5개의 삭제해야 할 중복 행이 존재함.

따라서 205 - 5 = 200 이 학생 수 라고 말할 수 있음
<br><br>

## Q02. 중복 행을 지우고 빈칸 채우기

`raw`는 그대로 두고 작업용 표 `clean`을 만듭니다. 아래 코드를 순서대로 완성하세요.

참고: N18·N20·N21

### Q02-1 · 복사본에서 중복 제거

아래 코드는 `raw`를 복사합니다. 이어서 **완전히 같은 행은 첫 번째만 남기고 나머지를 제거**한 결과를 `clean`에 저장하세요. 마지막에 `clean`의 행 수를 출력하세요.

프로그램이나 학년만 같다는 이유로 다른 학생의 행을 지우면 안 됩니다.

사용할 수 있는 기능: `drop_duplicates()`, `len()`

In [24]:
clean = raw.copy()
# 이어서 중복을 제거하고 결과를 clean에 저장하세요.
clean = clean.drop_duplicates()
# clean의 행 수를 출력하세요.
print("Num of clean row : ",len(clean))

Num of clean row :  200


### Q02-2 · 이동 시간의 중앙값 구하기

중복을 제거한 `clean`에서 `TravelMinutes`의 중앙값을 구해 **`median_time`**에 저장하고 출력하세요.

빈칸은 중앙값 계산에서 자동으로 제외됩니다.

사용할 수 있는 기능: `median()`

In [25]:
# 여기에 코드를 작성하세요.
median_time = clean["TravelMinutes"].median()
print("Median of TravelMinutes : ",median_time)

Median of TravelMinutes :  30.0


### Q02-3 · 이동 시간 빈칸 채우기

`clean`의 `TravelMinutes` 빈칸을 앞에서 구한 `median_time`으로 채우세요. 결과는 `clean`의 해당 열에 저장하고, 남은 빈칸 개수를 출력하세요.

이것은 과제에서 정한 대체 규칙이지, 실제 이동 시간을 알아내는 과정은 아닙니다.

사용할 수 있는 기능: `fillna()`, `isna()`, `sum()`

In [26]:
# 여기에 코드를 작성하세요.
print("Before : ",clean["TravelMinutes"].isna().sum())
clean["TravelMinutes"] = clean["TravelMinutes"].fillna(median_time)
print("After : ", clean["TravelMinutes"].isna().sum())

Before :  20
After :  0


### Q02-4 · 알게 된 경로의 빈칸 채우기

`clean`의 `Channel` 빈칸을 문자열 `Unknown`으로 채우세요. 결과는 `clean`의 해당 열에 저장하고, 남은 빈칸 개수를 출력하세요. `Unknown`은 ‘알 수 없음’이라는 뜻입니다.

사용할 수 있는 기능: `fillna()`

In [27]:
# 여기에 코드를 작성하세요.
print("Before : ",clean["Channel"].isna().sum())
clean["Channel"] = clean["Channel"].fillna("Unknown")
print("After : ",clean["Channel"].isna().sum())

Before :  12
After :  0


### Q02-5 · 원자료와 처리 결과 비교

**`clean`의 행 수**, **`clean`의 열별 빈칸 개수**, **`raw`의 열별 빈칸 개수**를 각각 출력하세요.

In [31]:
# 여기에 코드를 작성하세요.
print("Num of clean row : ",len(clean))
print("\n\nNum of None in each row of clean")
for attribute in clean.columns:
    print(attribute," : ",clean[attribute].isna().sum())
print("\n\nNum of None in each row of raw")
for attribute in raw.columns:
    print(attribute," : ",raw[attribute].isna().sum())

Num of clean row :  200


Num of None in each row of clean
RegistrationId  :  0
Program  :  0
Year  :  0
TravelMinutes  :  0
Channel  :  0
Attended  :  0


Num of None in each row of raw
RegistrationId  :  0
Program  :  0
Year  :  0
TravelMinutes  :  22
Channel  :  14
Attended  :  0


**Q02 · 결과 질문 1**

중복 제거 후 학생 수는 몇 명인가요? 이동 시간 빈칸에 채운 중앙값은 몇 분인가요? 앞에서 출력한 값을 적으세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

205명 -> 200명

이동 시간 median : 30.0 min
<br><br>

**Q02 · 결과 질문 2**

Q02-5에서 `clean`과 `raw`의 빈칸 개수가 어떻게 다른가요? **작업용 표를 바꿔도 원자료는 그대로 남았는지** 출력 결과를 근거로 설명하세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

raw는 정제되지 않은 데이터이기 때문에 결측지가 존재하고 총 36개이다.

clean은 None 값들을 fillna로 모두 채웠기 때문에 결측치가 0개이다.

clean = raw.copy()를 했기 때문에 깊은 복사로 메모리 공간이 달라서 value가 공유되지 않는다.
<br><br>

## Q03. 실제 참여한 학생 중 이동 시간이 긴 학생 찾기

Q02에서 만든 `clean`을 이어서 사용합니다.

참고: N16·N17

### Q03-1 · 두 조건에 맞는 학생 선택

**실제 참여했고 이동 시간이 40분 이상인 학생**을 선택해 `selected`에 저장하세요. 선택된 전체 학생 수를 출력하세요.

두 조건은 `Attended`가 `"o"`이고, `TravelMinutes`가 40 이상인 것입니다.

사용할 수 있는 기능: `loc`, `&`, `len()`

In [ ]:
# 여기에 코드를 작성하세요.
mask = 


### Q03-2 · 이동 시간이 긴 순서로 정렬

`selected`를 `TravelMinutes`가 큰 값부터 나오게 정렬하세요. 이동 시간이 같으면 `RegistrationId`가 작은 번호부터 나오게 하세요. 정렬한 결과를 **`ordered`**에 저장하세요.

사용할 수 있는 기능: `sort_values()`

In [ ]:
# 여기에 코드를 작성하세요.


### Q03-3 · 필요한 세 열의 처음 5행 출력

`ordered`에서 `RegistrationId`, `Program`, `TravelMinutes` 세 열만 선택하세요. 그 결과의 **처음 5행**을 `answer3`에 저장하고 출력하세요.

사용할 수 있는 기능: 열 이름 리스트 선택, `head()`

In [ ]:
# 여기에 코드를 작성하세요.


**Q03 · 결과 질문**

Q03-1의 조건을 만족한 학생은 모두 몇 명인가요? Q03-3에서 출력한 첫 학생의 신청번호와 예상 이동 시간도 적으세요. **전체 인원수와 출력한 5행의 개수를 구분**하세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

<br><br>

## Q04. 프로그램별 신청자 수와 실제 참여율 구하기

같은 `clean`을 계속 사용합니다.

참고: N23·N26

### Q04-1 · 문자로 된 참여 여부를 숫자로 바꾸기

`Attended`의 `"o"`는 1, `"x"`는 0으로 바꾸어 새 열 **`AttendedCode`**에 저장하세요. 원래 `Attended` 열은 남겨 둡니다. 새 열에 빈칸이 생겼는지 개수를 출력하세요.

사용할 수 있는 기능: `map()`, `isna()`, `sum()`

In [ ]:
# 여기에 코드를 작성하세요.


### Q04-2 · 프로그램마다 개수·합계·평균 계산

`Program` 값이 같은 학생끼리 묶고, **`AttendedCode`를 대상으로** 개수·합계·평균을 구하세요. 결과 표를 `summary`에 저장하고 전체를 출력하세요. 프로그램 하나당 한 행이 나와야 합니다.

결과 열 이름은 다음과 같이 지정하세요. 이 이름들은 원자료에 있던 열이 아니라, 계산 결과에 붙이는 이름입니다.

| 결과 열 이름 | 구할 내용 |
|---|---|
| applicants | 신청한 학생 수. 미참여 학생도 포함 |
| attendees | 실제 참여한 학생 수 |
| attendance_rate | 실제 참여한 학생 수 ÷ 신청한 학생 수. 0~1 사이의 비율 |

사용할 수 있는 기능: `groupby()`, `agg()`, `size`, `sum`, `mean`

In [ ]:
# 여기에 코드를 작성하세요.


**Q04 · 결과 질문 1**

참여 여부를 1과 0으로 바꿨습니다. 이 값의 **개수를 세면 왜 신청자 수**, **합계를 구하면 왜 참여자 수**가 되나요? 각각 한 문장으로 설명하세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

<br><br>

**Q04 · 결과 질문 2**

`AttendedCode`의 **평균**을 구하면 왜 참여율이 되나요? 평균의 분자와 분모가 무엇인지 생각해서 한 문장으로 설명하세요.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

<br><br>

## Q05. 어떤 프로그램을 늘리고 싶은지 제안하기

Q04에서 만든 `summary`를 사용합니다. 신청자 수·참여자 수·참여율은 모두 이 가상 데이터 안의 결과입니다.

참고: N17·N26

### Q05-1 · 실제 참여자가 많은 순서로 보기

`summary`를 **`attendees`가 큰 순서**로 정렬해서 출력하세요. 실제로 온 학생이 많은 프로그램부터 보려는 것입니다.

In [ ]:
# 여기에 코드를 작성하세요.


### Q05-2 · 참여율이 높은 순서로 보기

`summary`를 **`attendance_rate`가 큰 순서**로 정렬해서 출력하세요. 신청자 중 실제로 온 비율이 높은 프로그램부터 보려는 것입니다.

In [ ]:
# 여기에 코드를 작성하세요.


**Q05 · 결과 질문 1**

실제 참여자 수 1위 프로그램과 인원수, 참여율 1위 프로그램과 참여율을 각각 적으세요. **두 기준의 1위 프로그램은 같은가요?**

비율이 0.75이면 75%라고 쓰면 됩니다. 동점이 있으면 공동 1위로 적어도 됩니다.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

<br><br>

**Q05 · 결과 질문 2**

다음 축제에서 더 많은 학생을 받을 수 있도록 **규모를 늘리고 싶은 프로그램 하나**를 고르세요. `summary`의 수치 **두 개 이상**을 근거로 들어 3~5문장으로 제안하세요.

어느 프로그램을 골라도 수치와 이유가 연결되면 됩니다. 꼭 참여율만 중요한 것은 아닐 수도 있습니다.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

<br><br>

**Q05 · 결과 질문 3**

실제로 규모를 늘릴지 결정하기 전에 **추가로 알고 싶은 정보 한 가지와 필요한 이유**를 적으세요. 예를 들어 운영비·공간 크기·참가자 만족도는 현재 데이터에 없습니다.

**답변**

(이 셀을 더블클릭하여 답변을 작성하세요.)

<br><br>